content/courses/ada_2022_whats_new/chapters/big_numbers.rst

```diff
@@ -0,0 +1,165 @@
+intermediate result by :ada:`mod n` during mᵈ calculation then it will
+work. Let's write this operation as a function:
+
+.. code-block:: ada
+
+   --  Calculate M ** D mod N
+
+   function Power_Mod (M, D, N : Big_Integer) return Big_Integer is
+
+      function Is_Odd (X : Big_Integer) return Boolean is
+        (X mod 2 /= 0);
+
+      Result : Big_Integer := 1;
+      Exp    : Big_Integer := D;
+      Mult   : Big_Integer := M mod N;
+   begin
+      while Exp /= 0 loop
+         --  Loop invariant is Power_Mod'Result = Result * Mult**Exp mod N
+         if Is_Odd (Exp) then
+            Result := (Result * Mult) mod N;
+         end if;
+
+         Mult := Mult ** 2 mod N;
+         Exp := Exp / 2;
+      end loop;
+
+      return Result;
+   end Power_Mod;
+
+Let's check this with the example from
+`Wikipedia <https://en.wikipedia.org/wiki/RSA_(cryptosystem)>`_.
+In the example the `public key` is (n = 3233, e = 17) and the message
+is m = 65. The encrypted message is
+mᵉ mod n = 65¹⁷ mod 3233 = 2790 = c.
+
+.. code-block:: ada
+
+   Ada.Text_IO.Put_Line (Power_Mod (M => 65, D => 17, N => 3233)'Image);
+
+.. code-block::
+
+   2790
+
+To decrypt it with the public key (n = 3233, d = 413) we need to
+calculate cᵈ mod n = 2790⁴¹³ mod 3233
+
+.. code-block:: ada
+
+   Ada.Text_IO.Put_Line (Power_Mod (M => 2790, D => 413, N => 3233)'Image);
+
+.. code-block::
+
+   65
+
+So :ada:`65` is the original message m. Easy!
+
+Complete code snippet:
+
+.. code:: ada run_button project=Courses.Ada_2022_Whats_New.Big_Numbers_Tiny_RSA
+
+   pragma Ada_2022;
+
+   with Ada.Text_IO;
+   with Ada.Numerics.Big_Numbers.Big_Integers;
+   use  Ada.Numerics.Big_Numbers.Big_Integers;
+
+   procedure Main is
+
+      --  Calculate M ** D mod N
+
+      function Power_Mod (M, D, N : Big_Integer) return Big_Integer is
+
+         function Is_Odd (X : Big_Integer) return Boolean is
+            (X mod 2 /= 0);
+
+            Result : Big_Integer := 1;
+            Exp    : Big_Integer := D;
+            Mult   : Big_Integer := M mod N;
+         begin
+            while Exp /= 0 loop
+               --  Loop invariant is Power_Mod'Result = Result * Mult**Exp mod N
+               if Is_Odd (Exp) then
+                  Result := (Result * Mult) mod N;
+               end if;
+
+               Mult := Mult ** 2 mod N;
+               Exp := Exp / 2;
+            end loop;
+
+            return Result;
+         end Power_Mod;
+
+   begin
+      Ada.Text_IO.Put_Line (Big_Integer'Image(2 ** 256));
+      --  Encrypt:
+      Ada.Text_IO.Put_Line (Power_Mod (M => 65, D => 17, N => 3233)'Image);
+      --  Decrypt:
+      Ada.Text_IO.Put_Line (Power_Mod (M => 2790, D => 413, N => 3233)'Image);
+   end Main;
+
+Big Reals
+---------
+
+Besides Big_Integer, Ada 2022 provides `Big Reals`_.
+
+.. _`Big Reals`: http://www.ada-auth.org/standards/2xaarm/html/AA-A-5-7.html
+
+.. note::
+
+   Don't use Big_Numbers for cryptography because it is vulnerable to
```

Comment:
If Big_Numbers are not supposed to be used for this kind of application, shouldn't we have a different code example above with an application where it's safe to use Big_Numbers?